certora/specs/bridge.spec

```diff
@@ -0,0 +1,526 @@
+
+    uint256 balanceU1 = tokenBalanceOf(eB, asset, eB.msg.sender);
+    uint256 balanceA1 = tokenBalanceOf(eB, Atoken, eB.msg.sender);
+    uint256 balanceS1 = tokenBalanceOf(eB, static, eB.msg.sender);
+        uint256 staticAmount = deposit(eB, Atoken, l2Recipient, amount, referralCode, fromUA);
+    //uint256 balanceU2 = tokenBalanceOf(eB, asset, eB.msg.sender);
+    //uint256 balanceA2 = tokenBalanceOf(eB, Atoken, eB.msg.sender);
+    //uint256 balanceS2 = tokenBalanceOf(eB, static, eB.msg.sender);
+        initiateWithdraw_L2(eF, Atoken, staticAmount, eF.msg.sender, toUA);
+    uint256 balanceU3 = tokenBalanceOf(eF, asset, eF.msg.sender);
+    uint256 balanceA3 = tokenBalanceOf(eF, Atoken, eF.msg.sender);
+    uint256 balanceS3 = tokenBalanceOf(eF, static, eF.msg.sender);
+    
+    assert balanceS1 == balanceS3;
+    assert 
+        (balanceA1 == balanceA3 && balanceU1 == balanceU3);
+}
+
+// Checks that the transitions between static to dynamic are inverses.
+// Verified
+rule dynamicToStaticInversible1(uint256 amount)
+{
+    // We assume both indexes (L1,L2) are represented in Ray (1e27).
+    address asset;
+    requireRayIndex(asset);
+    uint256 dynm = _staticToDynamicAmount_Wrapper(amount, asset, LENDINGPOOL_L1);
+    uint256 stat = _dynamicToStaticAmount_Wrapper(dynm, asset, LENDINGPOOL_L1);
+    assert amount == stat;
+}
+
+// Checks that it isn't possible to gain from transforming dynamic to static
+// and back.
+// This is violated because the mul and div are not inverses of each other,
+// therefore can lead to mul(div(a,b),b) > a (depends on remainder value).
+rule dynamicToStaticInversible2(uint256 amount)
+{
+    // We assume both indexes (L1,L2) are represented in Ray (1e27).
+    address asset;
+    requireRayIndex(asset);
+    // Just for debugging.
+    uint256 indexL1 = LENDINGPOOL_L1.liquidityIndexByAsset(asset);
+    uint256 stat = _dynamicToStaticAmount_Wrapper(amount, asset, LENDINGPOOL_L1);
+    uint256 dynm = _staticToDynamicAmount_Wrapper(stat, asset, LENDINGPOOL_L1);
+    assert amount == dynm;
+    assert amount >= dynm;
+}
+
+// We make sure that the message sent booleans are always false,
+// meaning that, according to our implementation, no external call can
+// invoke receiveRewards and withdraw in Bridge.sol, but only through the 
+// designated functions in the harnessed Bridge contract.
+invariant alwaysUnSent(env e)
+   !withdrawMessageStatus(e) && !bridgeRewardsMessageStatus(e)
+
+// Check consistency of 'asset' being registered as the underlying
+// token of 'AToken', both in the AToken contract, and also in the 
+// mapping _aTokenData.
+// We exclude the 'initialize' function since it is called only once
+// in the code. 
+invariant underlying2ATokenConsistency(address AToken, address asset)
+     (asset !=0 <=> AToken !=0) 
+     =>
+     (getUnderlyingAssetHelper(AToken) == asset 
+     <=>
+     getUnderlyingAssetOfAToken(AToken) == asset)
+     filtered{f-> excludeInitialize(f)}
```

Comment:
what are we checking here? that the linkage is correct? is there a bug that you can insert to the code and this rule will fail?

Reply:
This invariant is part of the setup (not explicitly), so I just wanted to check that there is consistency between the AToken contract value and the one in the bridge.
If someone were to write a flawed code, this would have been broken.